### PR TITLE
Deprecate invoke and invoke_no_args in favor of invoke_batch

### PR DESCRIPTION
**Pull request by findepi**
`invoke_batch` covers all needs, so let's deprecate and eventually remove the redundant variants.

- https://github.com/apache/datafusion/pull/13179 deprecated functions in ScalarUDF, this PR deprecates equivalent functions in ScalarUDFImpl trait.

- follows https://github.com/apache/datafusion/pull/13001
- follows https://github.com/apache/datafusion/pull/13179

### CODE REVIEW

**Comment by findepi** [comment]
This requires lots of changes, so let's do this incrementally.

**Comment by findepi** [comment]
https://github.com/apache/datafusion/pull/13179 deprecated functions in ScalarUDF, this PR deprecates equivalent functions in ScalarUDFImpl trait.

@comphead please take a look

File: datafusion/expr/src/udf.rs [hunk 1]
```diff
@@ -489,19 +493,40 @@ pub trait ScalarUDFImpl: Debug + Send + Sync {
 
     /// Invoke the function with `args` and the number of rows,
     /// returning the appropriate result.
+    ///
+    /// The function will be invoked with the slice of [`ColumnarValue`]
```
Comment (alamb) [inline datafusion/expr/src/udf.rs:497]: ❤️

**Review by alamb — APPROVED** [review]
Thank you @findepi  -- this looks great.

I think we could file a ticket (good first issue) for migrating the rest of the built in functions (removing the #allow_deprecated)

**Comment by findepi** [comment]
> I think we could file a ticket (good first issue) for migrating the rest of the built in functions (removing the #allow_deprecated)

https://github.com/apache/datafusion/issues/13238

please add  a label

**Comment by findepi** [comment]
Pushed a commit fixing logical conflict with https://github.com/apache/datafusion/pull/13130 that was merged recently.

**Comment by findepi** [comment]
cc @comphead

**Comment by alamb** [comment]
Looks good -- thanks @findepi

**Pull request merged**